aspnetcore/blazor/forms-validation.md

```diff
@@ -223,12 +230,383 @@ In the following example:
+    <DataAnnotationsValidator />
+    <ValidationSummary />
+
+    ...
+
+</EditForm>
+
+@code {
+    private Starship starship = new Starship() { ProductionDate = DateTime.UtcNow };
+    private List<string> validationMessages;
+
+    private void HandleValidSubmit()
+    {
+        if (Validate(starship, out validationMessages))
+        {
+            // Process the form
+        }
+    }
+
+    private bool Validate(Starship starship, out List<string> messages)
+    {
+        messages = new List<string>();
+
+        if (starship.Classification == "Defense" &&
+            string.IsNullOrEmpty(starship.Description))
+        {
+            messages.Add("For a 'Defense' ship classification, " +
+                "'Description' is required.");
+
+            return false;
+        }
+
+        return true;
+    }
+}
+```
+
+### Use a custom business logic validation component
+
+An approach that takes advantage of the forms infrastructure in the framework is to:
+
+* Process model validation in the component.
+* Pass field validation errors in a dictionary to a custom validation component.
+
+Create a business logic validation component to display error messages returned by the server API (`BusinessValidator.cs`):
+
+```csharp
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.Forms;
+
+namespace {SOLUTION NAMESPACE}.Client
+{
+    public class BusinessValidator : ComponentBase
+    {
+        private ValidationMessageStore messageStore;
+
+        [CascadingParameter]
+        private EditContext CurrentEditContext { get; set; }
+
+        protected override void OnInitialized()
+        {
+            if (CurrentEditContext == null)
+            {
+                throw new InvalidOperationException(
+                    $"{nameof(BusinessValidator)} requires a cascading " +
+                    $"parameter of type {nameof(EditContext)}. " +
+                    $"For example, you can use {nameof(BusinessValidator)} " +
+                    $"inside an {nameof(EditForm)}.");
+            }
+
+            messageStore = new ValidationMessageStore(CurrentEditContext);
+
+            CurrentEditContext.OnValidationRequested += (s, e) => 
+                messageStore.Clear();
+            CurrentEditContext.OnFieldChanged += (s, e) => 
+                messageStore.Clear(e.FieldIdentifier);
+        }
+
+        public void DisplayErrors(Dictionary<string, List<string>> errors)
+        {
+            foreach (var err in errors)
+            {
+                messageStore.Add(CurrentEditContext.Field(err.Key), err.Value);
+            }
+
+            CurrentEditContext.NotifyValidationStateChanged();
+        }
+    }
+}
+```
+
+When validation messages are set in the component, they're added to the form's <xref:Microsoft.AspNetCore.Components.Forms.ValidationMessageStore> and shown by the `BusinessValidator` component:
```

Comment:
Not really, they are shown by the EditForm next to the field if there is a field with a matching key and (if present) in the ValidationSummary